File: back/src/service/authService.ts
import {retrieveUser, saveUser} from '../repository/userRepository';
import {genSalt, hash, compare} from 'bcrypt';
import {IUser, User} from '../models/user';
import {generateAccessToken} from '../middleware/auth';

const EMAIL_REGEX: RegExp = /^(?=.{1,254}$)(?=.{1,64}@)[-!#$%&'*+/0-9=?A-Z^_`a-z{|}~]+(\.[-!#$%&'*+/0-9=?A-Z^_`a-z{|}~]+)*@[A-Za-z0-9]([A-Za-z0-9-]{0,61}[A-Za-z0-9])?(\.[A-Za-z0-9]([A-Za-z0-9-]{0,61}[A-Za-z0-9])?)*$/;
const PASSWORD_CHARACTERS: string = '@?:-_~&=+/*%*^\\!.,;<>"\'()[]{}$';
const PASSWORD_LENGTH: number = 8;

type Credentials = {
  email: string;
  password: string;
}

const isString = (object: string | object): object is string => {
  return typeof object === 'string';
};

function reject(reason: string | Error): Promise<never> {
  if (isString(reason))
    return Promise.reject(new Error(reason));
  return Promise.reject(reason);
}

async function hashPassword(password: string): Promise<string> {
  const salt = await genSalt(10);

  return await hash(password, salt);
}

async function userExists(email: string): Promise<boolean> {
  return await retrieveUser(email)
    .then(() => {
      return true;
    }).catch(() => {
      return false;
    });
}

function isValidCredentials(credentials: Credentials): boolean {
  const email = credentials.email.trim();
  const password = credentials.password.trim();

  const validEmail: boolean = EMAIL_REGEX.test(email);
  const validPasswordLength: boolean = password.length >= PASSWORD_LENGTH;
  const specialCharacters: string[] = password.split('')
    .filter(character => PASSWORD_CHARACTERS.includes(character));

  return (validEmail && validPasswordLength && specialCharacters.length > 0);
}

async function register(credentials: Credentials): Promise<IUser> {
  const exists = await userExists(credentials.email);

  if (exists)
    return reject('User with this email already exists');
  if (!isValidCredentials(credentials))
    return reject('Invalid credentials');
  return saveUser({
    email: credentials.email,
    password: await hashPassword(credentials.password),
    password_length: credentials.password.length
  });
}

async function login(credentials: Credentials): Promise<string> {
  return retrieveUser(credentials.email)
    .then(async (user: IUser) => {
      const validCredentials = await compare(credentials.password, user.password);

      if (!validCredentials)
        return reject('Invalid credentials');
      return generateAccessToken(user);
    })
    .catch(reason => {
      return reject(reason);
    });
}

export type {Credentials};
export {isString, register, login, hashPassword};

File: back/src/main/java/com/javaspring/back/authentification/AuthService.java
package com.javaspring.back.authentification;


import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;


@Service
public class AuthService implements UserDetailsService {

    private final UserService userService;
    private final JwtUtil jwtUtil;
    private final BCryptPasswordEncoder passwordEncoder;

    private final String EMAIL_REGEX = "^(?=.{1,254}$)(?=.{1,64}@)[-!#$%&'*+/0-9=?A-Z^_`a-z{|}~]+(\\.[-!#$%&'*+/0-9=?A-Z^_`a-z{|}~]+)*@[A-Za-z0-9]([A-Za-z0-9-]{0,61}[A-Za-z0-9])?(\\.[A-Za-z0-9]([A-Za-z0-9-]{0,61}[A-Za-z0-9])?)*$";

    public AuthService(UserService userService, JwtUtil jwtUtil, BCryptPasswordEncoder passwordEncoder) {
        this.userService = userService;
        this.jwtUtil = jwtUtil;
        this.passwordEncoder = passwordEncoder;
    }

    @Override
    public AuthUserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
        User user = userService.retrieveUser(email);
        return new AuthUserDetails(user);
    }

    public void register(RegisterData credentials) {
        if (userService.userExists(credentials.getEmail())) {
            throw new RuntimeException("User with this email already exists");
        }

        if (!isValidCredentials(credentials)) {
            throw new RuntimeException("Invalid credentials");
        }

        User user = new User();
        user.setEmail(credentials.getEmail());
        user.setPassword(hashPassword(credentials.getPassword()));
        user.setBirthdate(credentials.getBirthdate());
        userService.saveUser(user);
    }

    public String login(Credentials credentials) {
        User user = userService.retrieveUser(credentials.getEmail());
        if (!isPasswordValid(credentials.getPassword(), user.getPassword())) {
            throw new RuntimeException("Invalid credentials");
        }
        return jwtUtil.generateToken(credentials.getEmail());
    }

    private boolean isValidCredentials(RegisterData credentials) {
        String email = credentials.getEmail().trim();
        String password = credentials.getPassword().trim();

        boolean validEmail = email.matches(EMAIL_REGEX);
        boolean validPasswordLength = password.length() >= 8;
        long specialCharactersCount = password.chars().filter(ch -> "@?:-_~&=+/*%*^\\!.,;<>\'\"()[]{}$".indexOf(ch) >= 0).count();

        return validEmail && validPasswordLength && specialCharactersCount > 0;
    }

    private boolean isPasswordValid(String rawPassword, String hashedPassword) {
        return passwordEncoder.matches(rawPassword, hashedPassword);
    }

    private String hashPassword(String password) {
        return passwordEncoder.encode(password);
    }
}
